### PR TITLE
feat: notification show feedback

### PR DESCRIPTION
**Pull request by marclucraft**
Add Success / Error notice in Post editor screen for a Notification send.

- Provides visible feedback, as customers mention not knowing if a notification has been sent, scheduled, or failed.
- Links to created notification in the OneSignal dashboard (opens in a new tab).

image
image
image

Related intercom:
https://app.intercom.com/a/inbox/344a89aeac3f033e4dec4370781543b948aece6d/inbox/admin/5709101/conversation/215473037560885?view=List

### CODE REVIEW

**Comment by sherwinski** [comment]
Thanks @marclucraft! Will take a look and test this sometime tomorrow or the next day.

**Comment by marclucraft** [comment]
Hey @sherwinski! Bumping this one 🤜🏼 🤛🏼

File: v3/onesignal-block-editor-notice.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+  var wasSaving = false;
+
+  var unsubscribe = subscribe(function () {
+    var editor = select("core/editor");
+    if (!editor) return;
+
+    var isSaving = editor.isSavingPost();
+    var isAutosave = editor.isAutosavingPost();
+    var isSavingNonAutosave = isSaving && !isAutosave;
```
Comment (sherwinski) [inline v3/onesignal-block-editor-notice.js]: I think it would be also worth gating on `isSavingMetaBoxes()` here because `isSavingPost()` could flip back to `false` before the metabox action completes.

File: v3/onesignal-notification.php [hunk 2]
```diff
@@ -121,19 +130,175 @@ function onesignal_create_notification($post, $notification_options = array())
...
+    );
+}
+
+/**
+ * AJAX handler that returns and clears the pending notice for the current user.
+ * Used by the block editor JS after a save completes.
+ */
+function onesignal_ajax_get_send_notice()
+{
+    check_ajax_referer('onesignal_notice_nonce', 'nonce');
```
Comment (sherwinski) [inline v3/onesignal-notification.php:271]: We should add a user capability check here

```suggestion
function onesignal_ajax_get_send_notice()
{
    if(!current_user_can('edit_post', $post_id)) wp_send_json_success(null);
    check_ajax_referer('onesignal_notice_nonce', 'nonce');
```

File: v3/onesignal-notification.php [hunk 3]
```diff
@@ -121,19 +130,175 @@ function onesignal_create_notification($post, $notification_options = array())
...
+    $notification_id = $notice['detail'] ?? '';
+    $dashboard_url   = (!empty($app_id) && !empty($notification_id))
+        ? 'https://dashboard.onesignal.com/apps/' . rawurlencode($app_id) . '/push/' . rawurlencode($notification_id)
+        : '';
+
+    $link = !empty($dashboard_url)
+        ? ' <a href="' . esc_url($dashboard_url) . '" target="_blank" rel="noopener noreferrer">'
+            . esc_html__('View the notification in the OneSignal Dashboard.', 'onesignal')
+            . '</a>'
+        : '';
```
Comment (sherwinski) [inline v3/onesignal-notification.php]: nit: it would be a little more readable to dedupe this and the `$scheduled_link` definition into a separate function instead of inlining them.

File: v3/onesignal-notification.php [hunk 4]
```diff
@@ -121,19 +130,175 @@ function onesignal_create_notification($post, $notification_options = array())
...
+{
+    $key = 'onesignal_send_notice_' . get_current_user_id() . '_' . $post_id;
+    set_transient($key, array('status' => $status, 'detail' => $detail), 60);
+}
+
+/**
+ * Reads the stored transient and renders a WP admin notice, then deletes it.
+ */
+function onesignal_display_send_notice()
+{
```
Comment (sherwinski) [inline v3/onesignal-notification.php:196]: nit: consider gating this on `is_admin()` since it currently loads on every admin request

**Review by sherwinski — CHANGES_REQUESTED** [review]
Looks good so far @marclucraft. Could we also add some tests before merging? At minimum a test per status branch verifying which `$status`/`$detail` should be stored.

File: v3/onesignal-block-editor-notice.js [hunk 5]
```diff
@@ -0,0 +1,111 @@
+(function () {
+  var subscribe = wp.data.subscribe;
+  var select = wp.data.select;
+  var dispatch = wp.data.dispatch;
+
+  var wasSaving = false;
+
+  var unsubscribe = subscribe(function () {
```
Comment (sherwinski) [inline v3/onesignal-block-editor-notice.js]: Actually I just noticed that this doesn't seem to get called anywhere?

File: v3/onesignal-block-editor-notice.js [hunk 6]
```diff
@@ -0,0 +1,111 @@
+(function () {
```
Comment (sherwinski) [inline v3/onesignal-block-editor-notice.js:1]: Let's update all of the `var` to be `let` in this file instead.

File: v3/onesignal-notification.php [hunk 7]
```diff
@@ -121,19 +130,175 @@ function onesignal_create_notification($post, $notification_options = array())
...
+        onesignal_store_send_notice('error', $response->get_error_message(), $post->ID);
     } else {
         // Save the notification ID for potential future cancellation
         $response_code = wp_remote_retrieve_response_code($response);
         if ($response_code === 200) {
             $response_body = wp_remote_retrieve_body($response);
             $response_data = json_decode($response_body, true);
-            if (!empty($response_data['id'])) {
-                onesignal_save_notification_id($post->ID, $response_data['id']);
+            $notification_id = $response_data['id'] ?? '';
```
Comment (sherwinski) [inline v3/onesignal-notification.php:140]: It looks like we're reusing this var for multiple purposes: it either holds a notification id for success/scheduled branches, or an error string for warning/error. I think renaming it would help future maintainers from potentially getting confused.

My suggestion would be to rename `$notification_id` → `$detail`, and only compute `$dashboard_url` inside the success/scheduled arms. Alternatively, store the id and the message under different keys (e.g. `notification_id`, `error_message`) so the variable is a bit self-describing.

File: v3/onesignal-block-editor-notice.js [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+      credentials: "same-origin",
+    })
+      .then(function (r) {
+        return r.json();
+      })
+      .then(function (response) {
+        if (!response.success || !response.data) return;
+
+        var status = response.data.status;
+        var notificationId = response.data.detail;
```
Comment (sherwinski) [inline v3/onesignal-block-editor-notice.js]: Same as the previous comment with respect to renaming this variable depending on which branch we're using it under.

File: v3/onesignal-notification.php [hunk 9]
```diff
@@ -121,19 +130,182 @@ function onesignal_create_notification($post, $notification_options = array())
...
+/**
+ * Builds a dashboard link anchor tag for use in admin notices.
+ */
+function onesignal_build_dashboard_link($url, $label)
+{
+    if (empty($url)) return '';
+    return ' <a href="' . esc_url($url) . '" target="_blank" rel="noopener noreferrer">'
+        . esc_html__($label, 'onesignal')
+        . '</a>';
+}
```
[the remark this answers is not in the document]
Comment (marclucraft) [reply v3/onesignal-notification.php:190]: i18n not supported right now

**Comment by sherwinski** [comment]
Going to test this out locally before merging

**Review by sherwinski — APPROVED** [review]
LGTM and great addition, thanks @marclucraft!
